### PR TITLE
Make command keys (:/?) configurable #672

### PR DESCRIPTION
**Pull request by nanjekyejoannah**

### CODE REVIEW

**Comment by The-Compiler** [comment]
The bindings you added will simply try to run nothing (or run `:search` without an argument). You should use the [`set-cmd-text` command](https://github.com/The-Compiler/qutebrowser/blob/master/doc/help/commands.asciidoc#set-cmd-text) instead to set the commandline text to `:`, `/` and `?`.

**Comment by The-Compiler** [comment]
Also, as mentioned in https://github.com/The-Compiler/qutebrowser/issues/672#issuecomment-239103814 you should remove the hardcoded checks in `keyinput.modeparsers.NormalKeyParser._handle_single_key`.

File: qutebrowser/keyinput/modeparsers.py [hunk 1]
```diff
@@ -29,9 +29,11 @@
 from qutebrowser.config import config
 from qutebrowser.keyinput import keyparser
 from qutebrowser.utils import usertypes, log, objreg, utils
+from qutebrowser.config.parsers import keyconf
 
 
-STARTCHARS = ":/?"
+conf = keyconf.KeyConfigParser(None, None) 
+STARTCHARS = conf.get_bindings_for('normal').keys()
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeparsers.py]: I'm not sure if this is going to work - I think it's fine to keep them hardcoded here, as the statusbar can still show `:`, `/` and `?` even if the key is rebound.

File: qutebrowser/config/configdata.py [hunk 2]
```diff
@@ -1458,6 +1458,9 @@ def data(readonly=False):
 
     ('normal', collections.OrderedDict([
         ('clear-keychain ;; search', ['<Escape>']),
+        ('set-cmd-text -s :search ', ['/']),
+        ('set-cmd-text -s :search --r', ['?']),
+        ('set-cmd-text -s : ', [':']),
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.py]: No need for the trailing spaces here, as they'll be stripped off while parsing the command (but the `-s` option to `set-cmd-text` will insert a space)

File: qutebrowser/keyinput/modeparsers.py [hunk 3]
```diff
@@ -29,8 +29,9 @@
 from qutebrowser.config import config
 from qutebrowser.keyinput import keyparser
 from qutebrowser.utils import usertypes, log, objreg, utils
+from qutebrowser.config.parsers import keyconf
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeparsers.py]: That import now isn't needed anymore

File: qutebrowser/keyinput/modeparsers.py [hunk 4]
```diff
@@ -29,8 +29,9 @@
...
 
-
+
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeparsers.py]: nitpick: You added a space here, please remove it again :wink:

File: qutebrowser/keyinput/modeparsers.py [hunk 5]
```diff
@@ -70,7 +71,7 @@ def _handle_single_key(self, e):
             self._debug_log("Ignoring key '{}', because the normal mode is "
                 "currently inhibited.".format(txt))
             return self.Match.none
-        if not self._keystring and any(txt == c for c in STARTCHARS):
+        if not self._keystring:
             message.set_cmd_text(self._win_id, txt)
             return self.Match.definitive
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeparsers.py]: This now causes qutebrowser to handle *any* keypress as an initial command char, which makes it crash when you press any key - you can get rid of this whole block.

File: qutebrowser/config/configdata.py [hunk 6]
```diff
@@ -1466,6 +1466,9 @@ def data(readonly=False):
         ('set-cmd-text :open -b -i {url:pretty}', ['xO']),
         ('set-cmd-text -s :open -w', ['wo']),
         ('set-cmd-text :open -w {url:pretty}', ['wO']),
+        ('set-cmd-text -s :search', ['/']),
+        ('set-cmd-text -s :search --r', ['?']),
```
Comment (The-Compiler) [inline qutebrowser/config/configdata.py]: Sorry, missed that with the last review: This should be `-r` and not `--r` (single letter options have single dashes)

File: qutebrowser/keyinput/modeparsers.py [hunk 7]
```diff
@@ -231,7 +228,7 @@ def execute(self, cmdstr, keytype, count=None):
         if keytype == self.Type.chain:
             hintmanager = objreg.get('hintmanager', scope='tab',
                                      window=self._win_id, tab='current')
-            hintmanager.fire(cmdstr)
+            hintmanager.handle_partial_key(cmdstr)
```
Comment (The-Compiler) [inline qutebrowser/keyinput/modeparsers.py]: This seems like another change coming from `master` which shouldn't be in this branch?

**Comment by nanjekyejoannah** [comment]
I just Updated..

**Comment by The-Compiler** [comment]
Thanks! I squashed this and merged it as 056c0c3c14c2f0aad7603bcdb069e067ab7268dd, with some cleanups in 29cd878902b3b82d50542f9c185bf244bd8dbfaa (removing another space and a now unnecessary import) and d25fde4f29f8af5a1d2b1c26278c49e57723e1ab (somehow the change came back...).

**Comment by nanjekyejoannah** [comment]
Great thanks!!!

**Comment by mvucBmM0** [comment]
Thanks guys

**Pull request closed without merging**